### PR TITLE
Add proxy-wide alerts (supports Velocity & Bungeecord)

### PR DESCRIPTION
**Pull request by FlamWy**
#407

image

### CODE REVIEW

**Comment by iFlamingoz** [comment]
> Can this be made more proxy-agonistic? BungeeCord isn't the only proxy software that exists, and it looks like this PR is incompatible with velocity in multiple ways.

Velocity has bungee messaging support and I have tested this on velocity.

**Comment by Jaren8r** [comment]
1. This won't work with velocity modern forwarding as your code is only checking spigot.yml to see if bungeecord is set to true
2. ~~Velocity doesn't emit a PluginMessageEvent for plugin messages on the BungeeCord channel if `bungee-plugin-message-channel` is true, which is a default config option. A separate plugin messaging channel may be better for this.~~
3. ~~If this is supposed to work with multiple proxies, then why name it specifically bungeecord in the config file?~~

Whoops, I'm kind of an idiot. I only half-looked at it and I incorrectly assumed that you would need a companion plugin on your proxy to send the messages. I completely forgot you could use the BungeeCord message to forward messages to other servers.

File: src/main/java/ac/grim/grimac/manager/init/start/PacketManager.java [hunk 1]
```diff
@@ -26,13 +31,21 @@ public void start() {
...
         PacketEvents.getAPI().getEventManager().registerListener(new PacketPlayerSteer());
+        
 
         if (PacketEvents.getAPI().getServerManager().getVersion().isNewerThanOrEquals(ServerVersion.V_1_18)) {
             PacketEvents.getAPI().getEventManager().registerListener(new PacketWorldReaderEighteen());
         } else {
             PacketEvents.getAPI().getEventManager().registerListener(new BasePacketWorldReader());
         }
 
+        AlertPluginMessenger.setBungeeEnabled(YamlConfiguration.loadConfiguration(new File("spigot.yml")).getBoolean("settings.bungeecord"));
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/manager/init/start/PacketManager.java:42]: This isn't compatible with velocity modern forwarding setups.

File: src/main/java/ac/grim/grimac/events/packets/AlertPluginMessenger.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+public class AlertPluginMessenger extends PacketListenerAbstract {
+	@Getter
+	@Setter
+	private static boolean bungeeEnabled;
+	
+	public AlertPluginMessenger() {
+		GrimAPI.INSTANCE.getPlugin().getServer().getMessenger().registerOutgoingPluginChannel(GrimAPI.INSTANCE.getPlugin(), "BungeeCord");
+	}	
+	
+    @Override
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/events/packets/AlertPluginMessenger.java:35]: Indentation error

File: src/main/java/ac/grim/grimac/events/packets/AlertPluginMessenger.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        if (event.getPacketType() == PacketType.Play.Client.PLUGIN_MESSAGE) {
+            WrapperPlayClientPluginMessage packet = new WrapperPlayClientPluginMessage(event);
+            
+            if (!packet.getChannelName().equals("BungeeCord") || !AlertPluginMessenger.canReceiveAlerts()) return;
+            
+            ByteArrayDataInput in = ByteStreams.newDataInput(packet.getData());
+            
+            if (!in.readUTF().equals("GRIMAC")) return;
+            
+			byte[] msgbytes = new byte[in.readShort()];
```
Comment (Jaren8r) [inline src/main/java/ac/grim/grimac/events/packets/AlertPluginMessenger.java:46]: Indentation error

File: src/main/resources/config/en.yml [hunk 4]
```diff
@@ -9,6 +9,11 @@ alerts:
   enable-on-join: true
   # In addition to broadcasting alerts to players, should they also be sent to the console?
   print-to-console: true
+  bungeecord:
+     # Should alerts be sent to other servers connected to your proxy?
+     send: false
+     # Should alerts received by other servers be broadcasted?
```
Comment (Jaren8r) [inline src/main/resources/config/en.yml:15]:
```suggestion
     # Should alerts received from other servers be broadcasted?
```

**Review by Jaren8r — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/commands/GrimSendAlert.java [hunk 5]
```diff
@@ -19,6 +20,10 @@ public void sendAlert(String string) {
         for (Player bukkitPlayer : GrimAPI.INSTANCE.getAlertManager().getEnabledAlerts()) {
             bukkitPlayer.sendMessage(string);
         }
+        
+        if (ProxyAlertMessenger.canSendAlerts()) {
+        	ProxyAlertMessenger.sendPluginMessage(string);
```
Comment (NoJokeFNA) [inline src/main/java/ac/grim/grimac/commands/GrimSendAlert.java:25]: Add more spaces:
```java
        if (ProxyAlertMessenger.canSendAlerts()) {
            ProxyAlertMessenger.sendPluginMessage(string);
```

File: src/main/java/ac/grim/grimac/events/packets/ProxyAlertMessenger.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Override
+    public void onPacketReceive(final PacketReceiveEvent event) {
+        if (event.getPacketType() != PacketType.Play.Client.PLUGIN_MESSAGE || !ProxyAlertMessenger.canReceiveAlerts())
+            return;
+
+        WrapperPlayClientPluginMessage packet = new WrapperPlayClientPluginMessage(event);
+
+        String channelName = packet.getChannelName().toString();
+
```
Comment (NoJokeFNA) [inline src/main/java/ac/grim/grimac/events/packets/ProxyAlertMessenger.java:45]: Why soo many empty spaces? Remove the useless ones

**Review by NoJokeFNA — CHANGES_REQUESTED** [review]

**Comment by caoli5288** [comment]
Can it support multi-bungee?

**Comment by FlamWy** [comment]
> Can it support multi-bungee?

I am not sure.

**Pull request merged**